packages/expo-notifications/ios/EXNotifications/Notifications/Emitter/EXNotificationsEmitter.m

```diff
@@ -0,0 +1,85 @@
+// Copyright 2018-present 650 Industries. All rights reserved.
+
+#import <EXNotifications/EXNotificationsEmitter.h>
+#import <EXNotifications/EXNotificationSerializer.h>
+#import <EXNotifications/EXNotificationCenterDelegate.h>
+
+#import <UMCore/UMEventEmitterService.h>
+
+static NSString * const onDidReceiveNotification = @"onDidReceiveNotification";
+static NSString * const onDidReceiveNotificationResponse = @"onDidReceiveNotificationResponse";
+
+@interface EXNotificationsEmitter ()
+
+@property (nonatomic, weak) id<EXNotificationCenterDelegate> notificationCenterDelegate;
+
+@property (nonatomic, assign) BOOL isListening;
+@property (nonatomic, assign) BOOL isBeingObserved;
+
+@property (nonatomic, weak) id<UMEventEmitterService> eventEmitter;
+
+@end
+
+@implementation EXNotificationsEmitter
+
+UM_EXPORT_MODULE(ExpoNotificationsEmitter);
+
+# pragma mark - UMModuleRegistryConsumer
+
+- (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
+{
+  _eventEmitter = [moduleRegistry getModuleImplementingProtocol:@protocol(UMEventEmitterService)];
+  _notificationCenterDelegate = [moduleRegistry getSingletonModuleForName:@"NotificationCenterDelegate"];
+}
+
+# pragma mark - UMEventEmitter
+
+- (NSArray<NSString *> *)supportedEvents
+{
+  return @[onDidReceiveNotification, onDidReceiveNotificationResponse];
+}
+
+- (void)startObserving
+{
+  [self setIsBeingObserved:YES];
+}
+
+- (void)stopObserving
+{
+  [self setIsBeingObserved:NO];
+}
+
+- (void)setIsBeingObserved:(BOOL)isBeingObserved
+{
+  _isBeingObserved = isBeingObserved;
+  BOOL shouldListen = _isBeingObserved;
+  if (shouldListen && !_isListening) {
+    [_notificationCenterDelegate addDelegate:self];
+    _isListening = YES;
+  } else if (!shouldListen && _isListening) {
+    [_notificationCenterDelegate removeDelegate:self];
+    _isListening = NO;
+  }
+}
+
+# pragma mark - EXNotificationsDelegate
+
+- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo fetchCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler
+{
+  // Background task execution would happen here.
+  completionHandler(UIBackgroundFetchResultNoData);
+}
+
+- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler
+{
+  [_eventEmitter sendEventWithName:onDidReceiveNotificationResponse body:[EXNotificationSerializer serializedNotificationResponse:response]];
+  completionHandler();
+}
+
+- (void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions))completionHandler
+{
+  [_eventEmitter sendEventWithName:onDidReceiveNotification body:[EXNotificationSerializer serializedNotification:notification]];
+  completionHandler(UNNotificationPresentationOptionAlert);
```

Comment:
i'm pretty sure we currently allow at least a build-time configuration option to set this behavior to `UNNotificationPresentationOptionNone` (and that may even be the default). would be great to allow developers to at least choose between `None` and `Alert` (if not the other options).

Reply:
hmm, never mind, i see this is part of #6796 ! Could you clarify the relationship between these two modules? Is it that if developers want fine-grained control over how notifications are handled, they can use `NotificationHandler` but if they want simple/default behavior they can use `NotificationEmitter` instead? Does this module's delegate still get called (and therefore call `completionHandler(UNNotificationPresentationOptionAlert);`) even if you're using `NotificationHandler`?

Reply:
Yes, that's how I would imagine it — `NotificationEmitter` is a plain emitter, one could use it to log notifications, send them to server, respond to user input (like when the user enters a text in a notification) etc., whereas `NotificationHandler` would be responsible for responding to notification requests (let the system know what to do with it).

I'm open to changes in this architecture, it seemed simple to me, but maybe it may be too daunting for the end user. Maybe `NotificationHandler` should also subscribe to these events, so that the user would have one handler responsible for handling everything one would need to do?

Even if we would provide these simple listening functionalities in `NotificationHandler`, I still think exposing these no-action emitters would be beneficial for the future usages (sort of `MessageQueue.spy`-like, or eg. for analytics libraries, etc.)

Reply:
Gotcha. I think that architecture makes sense and will handle a lot of use cases simply while also providing more power and flexibility for developers who need it.

I was thinking it might be nice to have a simple option to just show system alerts for all notifications while your app is in the foreground, but the more I think about it, you would probably almost always want to implement a notification handler anyway. So imo this seems good. 👍